casper/src/main/scala/coop/rchain/casper/Validate.scala

```diff
@@ -141,6 +150,44 @@ object Validate {
       }
     }
 
+  /**
+    * Validate no deploy by the same (user, millisecond timestamp)
+    * has been produced in the chain
+    */
+  def repeatDeploy[F[_]: Monad: Log: BlockStore](block: BlockMessage,
+                                                 genesis: BlockMessage,
+                                                 dag: BlockDag): F[Boolean] =
+    BlockStore[F].asMap().flatMap { internalMap: Map[BlockHash, BlockMessage] =>
+      {
+        val deployKeySet = block.getBody.newCode.map(d => (d.getRaw.user, d.getRaw.timestamp)).toSet
+        val iterator = DagOperations.bfTraverse(Some(genesis))(x =>
+          childMapIterator(dag.childMap.getOrElse(x.blockHash, Set.empty), dag, internalMap))
+        val repeatedBlocks = iterator.filter(it => {
+          it.body.exists(
+            _.newCode.exists(
+              _.raw.exists(p => deployKeySet.contains((p.user, p.timestamp)))
+            )
+          )
+        })
+        // it is allowed for there to be the same (user, timestamp) transaction in both forks
+        val parents = ProtoUtil.parents(block)
+        val invalid = repeatedBlocks.exists(b1 => {
+          parents.contains(b1.blockHash)
+        })
```

Comment:
How does this cover forks that are longer than one block? Are you familiar with the syntax "P <- C" meaning block P is the parent of block C

A <- B1 <- C1 <- D1 <- E1
A <- B2 <- C2 <- D2<- E2

Imagine that B1 and D2 have the same (user, timestamp) transaction. That would be ok too.

Reply:
I think it works. When you deploy D2, I will firstly traverse the dag to find these blocks that contain deploy with same (user, timestamp) transaction, which is B1 in your case. But D2's parents is [A, B2, C2, D2], it doesn't contain B2, so it is valid.

Reply:
`ProtoUtil.parents(D2) == C2`. It doesn't include B2 or A. You'll probably need to write something similar to `ProtoUtil.getMainChain`. Make sure to add a test case that fails otherwise.